go/vt/mysqlctl/azblobbackupstorage/azblob.go

```diff
@@ -0,0 +1,297 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package azblobbackupstorage implements the BackupStorage interface
+// for Azure Blob Storage
+package azblobbackupstorage
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"io"
+	"net/url"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/Azure/azure-storage-blob-go/azblob"
+	"vitess.io/vitess/go/vt/concurrency"
+	"vitess.io/vitess/go/vt/log"
+	"vitess.io/vitess/go/vt/mysqlctl/backupstorage"
+)
+
+var (
+	// This is the account name
+	accountName = flag.String("azblob_backup_account_name", "", "Azure Account Name for Backups")
+
+	// This is the private access key
+	accountKey = flag.String("azblob_backup_account_key", "", "Azure Account Key")
+
+	// This is the name of the container that will store the backups
+	containerName = flag.String("azblob_backup_contain_name", "", "Azure Blob Contain Name")
+
+	// This is an optional previx to prepend to all files
+	prefix = flag.String("azblob_backup_prefix", "", "Azure Blob prefix")
+
+	azBlobParallelism = flag.Int("azblob_parallelism", 1, "Azure blob operation parallelism (requires extra memory when increased)")
+)
+
+const (
+	defaultRetryCount = 5
+	delimiter         = "/"
+)
+
+func azCredentials() (*azblob.SharedKeyCredential, error) {
+	return azblob.NewSharedKeyCredential(*accountName, *accountKey)
+}
+
+func azServiceURL(credentials azblob.Credential) azblob.ServiceURL {
+	pipeline := azblob.NewPipeline(credentials, azblob.PipelineOptions{
+		Retry: azblob.RetryOptions{
+			Policy:   azblob.RetryPolicyFixed,
+			MaxTries: defaultRetryCount,
+			// Per https://godoc.org/github.com/Azure/azure-storage-blob-go/azblob#RetryOptions
+			// This shuld be set to a very nigh number ( they claim 60s per MB ). That could end up being days so we are limiting this to four hours
+			TryTimeout: 4 * time.Hour,
+		},
+	})
+	u, _ := url.Parse(fmt.Sprintf("https://%s.blob.core.windows.net/", *accountName))
+	return azblob.NewServiceURL(*u, pipeline)
+}
+
+// AZBlobBackupHandle implements BackupHandle for Azure Blob service.
+type AZBlobBackupHandle struct {
+	bs        *AZBlobBackupStorage
+	dir       string
+	name      string
+	readOnly  bool
+	waitGroup sync.WaitGroup
+	errors    concurrency.AllErrorRecorder
+}
+
+// Directory implements BackupHandle.
+func (bh *AZBlobBackupHandle) Directory() string {
+	return bh.dir
+}
+
+// Name implements BackupHandle.
+func (bh *AZBlobBackupHandle) Name() string {
+	return bh.name
+}
+
+// AddFile implements BackupHandle.
+func (bh *AZBlobBackupHandle) AddFile(ctx context.Context, filename string, filesize int64) (io.WriteCloser, error) {
+	if bh.readOnly {
+		return nil, fmt.Errorf("AddFile cannot be called on read-only backup")
+	}
+	// Error out if the file size it too large ( ~4.75 TB)
+	if filesize > azblob.BlockBlobMaxStageBlockBytes*azblob.BlockBlobMaxBlocks {
```

Comment:
Note that by doing this here, you are limiting xtrabackup archives to be smaller than what Azure blob can support. That is because we pass in the total file size on disk, but there is optional compression on top of that.